userapi/src/main/java/fk/prof/userapi/model/tree/CalleesTreeView.java

```diff
@@ -0,0 +1,88 @@
+package fk.prof.userapi.model.tree;
+
+import fk.prof.aggregation.proto.AggregatedProfileModel.FrameNode;
+import fk.prof.userapi.cache.Cacheable;
+import fk.prof.userapi.model.ProfileView;
+import fk.prof.userapi.model.ProfileViewType;
+import fk.prof.userapi.model.Tree;
+import fk.prof.userapi.model.TreeView;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Created by gaurav.ashok on 01/06/17.
+ */
+public class CalleesTreeView implements TreeView<IndexedTreeNode<FrameNode>>, Cacheable<ProfileView> {
+
+    private Tree<FrameNode> callTree;
+    private List<Integer> hotMethodNodeIds;
+
+    public CalleesTreeView(Tree<FrameNode> callTree, List<Integer> hotMethodNodeIds) {
+        this.callTree = callTree;
+        this.hotMethodNodeIds = hotMethodNodeIds;
+    }
+
+    public List<Integer> getRoots() {
+        return hotMethodNodeIds;
+    }
+
+    public List<IndexedTreeNode<FrameNode>> getSubTrees(List<Integer> ids, int maxDepth, boolean forceExpand) {
+        return new Expander(ids, maxDepth, forceExpand).expand();
+    }
+
+    @Override
+    public ProfileViewType getType() {
+        return ProfileViewType.CALLEES;
+    }
+
+    /**
+     * This is a helper class in order to contain variables as field members
+     * which will remain constant in the expand method
+     */
+    private class Expander {
+        List<Integer> ids;
+        int maxDepth;
+        boolean forceExpand;
+
+        Expander(List<Integer> ids, int maxDepth, boolean forceExpand) {
+            this.ids = ids;
+            this.maxDepth = maxDepth;
+            this.forceExpand = forceExpand;
+        }
+
+        List<IndexedTreeNode<FrameNode>> expand() {
+            Map<String, List<IndexedTreeNode<FrameNode>>> idxGroupedByMethodIdLineNum = ids.stream()
+                .map(e -> new IndexedTreeNode<>(e, callTree.getNode(e)))
+                .collect(Collectors.groupingBy((IndexedTreeNode<FrameNode> e) -> String.valueOf(e.getData().getMethodId()) + ":" + String.valueOf(e.getData().getLineNo())));
+            return idxGroupedByMethodIdLineNum.values().stream().peek(this::expand).flatMap(List::stream).collect(Collectors.toList());
+        }
+
+        void expand(List<IndexedTreeNode<FrameNode>> nodes) {
```

Comment:
Have we considered de-duping (aggregating the tree) here itself? Strategies will obviously depend on work type but we can abstract out that logic.
Instead of `void expand(nodes)`, there can be a `node expandAndGroup(nodes)` method which takes aggregatable nodes as input, expands them and returns aggregated node.
Something like:
```
expandAndGroup(nodes, curDepth) {
    //id means method-id+line-num
    // nodes is list<nodes> where all entries belong to same method-id+line-num

    return if too deep
    aggregatedNode := aggregated input (list<nodes>) to get this
    groupedParents := Map<id, list<nodes>>
    for every (id,parents): aggregatedNode.addChild(expandAndGroup(parents, curDepth+1))
    return aggregatedNode
}
```

Reply:
i dont think in our case grouping was feasible. We discussed it before. The nodes that will be grouped by UI, have different properties(node id, sample count, etc). UI needs this information, for presentation and further getSubtree calls.